source/postprocess/gravity_point_values.cc

```diff
@@ -240,17 +150,210 @@ namespace aspect
+                                    + file_prefix);
+
+      // Get quadrature formula and increase the degree of quadrature over the velocity
+      // element degree.
+      const unsigned int degree = this->get_fe().base_element(this->introspection()
+                                                              .base_elements.velocities).degree
+                                  + quadrature_degree_increase;
+      const QGauss<dim> quadrature_formula (degree);
+      FEValues<dim> fe_values (this->get_mapping(),
+                               this->get_fe(),
+                               quadrature_formula,
+                               update_values   |
+                               update_gradients |
+                               update_quadrature_points |
+                               update_JxW_values);
+
+
+      // Get the value of the universal gravitational constant:
+      const double G = aspect::constants::big_g;
+
+      const unsigned int n_quadrature_points_per_cell = quadrature_formula.size();
+
+      const unsigned int n_satellites = satellite_positions_spherical.size();
+
+
+      // This is the main loop which computes gravity acceleration, potential and
+      // gradients at a point located at the spherical coordinate [r, phi, theta].
+      // This loop corresponds to the 3 integrals of Newton law:
+      std::vector<double>                 local_g_potential (n_satellites);
+      std::vector<Tensor<1,dim>>          local_g (n_satellites);
+      std::vector<Tensor<1,dim>>          local_g_anomaly (n_satellites);
+      std::vector<SymmetricTensor<2,dim>> local_g_gradient (n_satellites);
+
+      MaterialModel::MaterialModelInputs<dim> in(quadrature_formula.size(),
+                                                 this->n_compositional_fields());
+      MaterialModel::MaterialModelOutputs<dim> out(quadrature_formula.size(),
+                                                   this->n_compositional_fields());
+
+      for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+        if (cell->is_locally_owned())
+          {
+            // Evaluate the solution at the quadrature points of this cell
+            fe_values.reinit (cell);
+
+            in.reinit(fe_values, cell, this->introspection(), this->get_solution(), false);
+            this->get_material_model().evaluate(in, out);
+
+            for (unsigned int p=0; p < n_satellites; ++p)
+              {
+                const Point<dim> satellite_position = satellite_positions_cartesian[p];
+
+                for (unsigned int q = 0; q < n_quadrature_points_per_cell; ++q)
+                  {
+                    const Tensor<1,dim> r_vector = satellite_position - fe_values.quadrature_point(q);
+
+                    const double r_squared = r_vector.norm_square();
+                    const double r = std::sqrt(r_squared);
+
+                    const double density_JxW = out.densities[q] * fe_values.JxW(q);
```

Comment:
Did you measure if the new implementation is faster for a reasonably large number of satellites (say 1000 or so)? I think part of why the old implementation had this complicated split loop was in order to compute `density_JxW` exactly once per quadrature point. With your current implementation you compute it `n_satellites * n_quadrature_points` times. Is it possible to reorder the loops over satellites and quadrature points to avoid this?
I can do a complete review later, I have some ideas for smaller optimizations, but wanted to ask if you benchmarked it first.

Reply:
I could move this out, but the real cost is *computing* `JxW` and the densities. For `JxW`, for example, you have to evaluate the gradients of the shape functions of a Q4 mapping at the current quadrature point, multiply these gradients by the locations of the support points, and compute the determinant of the resulting matrix. All of this is going to cost hundreds of FP operations. Is your question whether the multiplications here are too expensive if one has a lot of satellites? I guess that could be -- I could just move this outside the loop over the satellites.

Reply:
I should probably also move `r_squared` and `r` out of the loop.

Reply:
Ah no, I can't, it depends on `q` :-(

Reply:
>  Is your question whether the multiplications here are too expensive if one has a lot of satellites? I guess that could be -- I could just move this outside the loop over the satellites.

Yes that was my question, please do move it out of the loop.